pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts

```diff
@@ -84,3 +92,73 @@ const deriveDatabaseDetails = (
     numIndexRecommendations,
   };
 };
+
+interface DerivedTableDetailsParams {
+  dbName: string;
+  tables: string[];
+  tableDetails: Record<string, TableDetailsState>;
+  nodeRegions: Record<string, string>;
+  isTenant: boolean;
+}
+
+export const deriveTableDetailsMemoized = createSelector(
+  (params: DerivedTableDetailsParams) => params.dbName,
+  (params: DerivedTableDetailsParams) => params.tables,
+  (params: DerivedTableDetailsParams) => params.tableDetails,
+  (params: DerivedTableDetailsParams) => params.nodeRegions,
+  (params: DerivedTableDetailsParams) => params.isTenant,
+  (
+    dbName,
+    tables,
+    tableDetails,
+    nodeRegions,
+    isTenant,
+  ): DatabaseDetailsPageDataTable[] => {
+    tables = tables || [];
+    return tables.map(table => {
+      const tableID = generateTableID(dbName, table);
+      const details = tableDetails[tableID];
+      return deriveDatabaseTableDetails(table, details, nodeRegions, isTenant);
+    });
+  },
+);
+
+const deriveDatabaseTableDetails = (
+  table: string,
+  details: TableDetailsState,
+  nodeRegions: Record<string, string>,
+  isTenant: boolean,
+): DatabaseDetailsPageDataTable => {
+  const results = details?.data?.results;
+  const grants = results?.grantsResp.grants || [];
+  const normalizedRoles = normalizeRoles(grants.map(grant => grant.user));
+  const normalizedPrivileges = normalizePrivileges(
+    [].concat(...grants.map(grant => grant.privileges)),
+  );
+  const nodes = results?.stats.replicaData.nodeIDs || [];
+  return {
+    name: table,
+    loading: !!details?.inFlight,
+    loaded: !!details?.valid,
```

Comment:
I'm curious why we need both of these.

Reply:
We check both fields to determine whether to refresh table details or not (i.e. if table is not loaded and not loading, we should fetch it's data).